File: gitdot-core/src/model/core/organization.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, Type};
use uuid::Uuid;

#[derive(Debug, Clone, FromRow)]
pub struct Organization {
    pub id: Uuid,
    pub name: String,

    pub display_name: Option<String>,
    pub location: Option<String>,
    pub readme: Option<String>,
    pub links: Vec<String>,

    pub created_at: DateTime<Utc>,

    #[sqlx(json(nullable))]
    pub members: Option<Vec<OrganizationMember>>,
}

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct OrganizationMember {
    pub id: Uuid,
    pub user_id: Uuid,
    pub user_name: String,

    pub role: OrganizationRole,
    pub role_description: Option<String>,

    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Eq, Type, Serialize, Deserialize)]
#[sqlx(type_name = "core.organization_role", rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum OrganizationRole {
    Admin,
    Member,
}

File: gitdot-web/packages/api/src/resource/organization.ts
import { z } from "zod";

export const OrganizationResource = z.object({
  id: z.uuid(),
  name: z.string(),
  created_at: z.iso.datetime(),
  readme: z.string().nullable().optional(),
  links: z.array(z.string()).default([]),
});
export type OrganizationResource = z.infer<typeof OrganizationResource>;

export const OrganizationMemberResource = z.object({
  id: z.uuid(),
  user_id: z.uuid(),
  organization_id: z.uuid(),
  role: z.string(),
  role_description: z.string().nullable().optional(),
  created_at: z.iso.datetime(),
  user_name: z.string(),
});
export type OrganizationMemberResource = z.infer<
  typeof OrganizationMemberResource
>;
